servo/connectors/wavefront.py

```diff
@@ -0,0 +1,353 @@
+import asyncio
+import datetime
+import functools
+import re
+from typing import Dict, Iterable, List, Optional, Tuple
+
+import httpcore._exceptions
+import httpx
+import pydantic
+
+import servo
+
+DEFAULT_BASE_URL = "http://wavefront.com:2878"
+API_PATH = "/api/v2/"
+
+
+class WavefrontMetric(servo.Metric):
+    """WavefrontMetric objects describe metrics that can be measured by querying
+    Wavefront.
+    """
+
+    query: str
+    """A WQL query that returns the value of the target metric.
+
+    For details on Wavefront, see the [Wavefront
+    Querying](https://docs.Wavefront.com/query_language_getting_started.html)
+    documentation.
+    """
+
+    granularity: pydantic.constr(
+        regex=r"^[smhd]"
+    ) = "m"
+
+    """The resolution of the query. The granularity resolution determines the number of data points captured across a
+    query range. A query's granularity is completely independent from any range durations specified in the WQL expression it evaluates.
+    
+    Available values: s, m, h, d.
+    """
+
+    summarization: pydantic.constr(
+        min_length=3,
+        max_length=6
+    ) = "LAST"
+    """Summarization strategy to use when bucketing points together.
+
+    Available values: MEAN, MEDIAN, MIN, MAX, SUM, COUNT, LAST, FIRST.
+    """
```

Comment:
We should turn this one into an enum instead of doing the pseudo-check with line lengths. Pydantic will constrain the values to members of the enumeration and if you also inherit from `str`, they can be interpolated directly into URL query strings, log messages, etc:

```python
class Summarization(str, enum.Enum):
    mean = "MEAN"
    median = "MEDIAN"
    min = "MIN"
    max = "MAX"
    sum = "SUM"
    count = "COUNT"
    last = "LAST"
    first = "FIRST"
```